src/ring_theory/polynomial.lean

```diff
@@ -47,6 +49,78 @@ begin
   apply le_trans (degree_X_pow_le _) (with_bot.coe_le_coe.2 $ nat.le_of_lt_succ $ finset.mem_range.1 hk)
 end
 
+def restriction (p : polynomial R) : polynomial (ring.closure (↑p.frange : set R)) :=
+⟨p.support, λ i, ⟨p.to_fun i,
+  if H : p.to_fun i = 0 then H.symm ▸ is_add_submonoid.zero_mem _
+  else ring.subset_closure $ finsupp.mem_frange.2 ⟨H, i, rfl⟩⟩,
+λ i, finsupp.mem_support_iff.trans (not_iff_not_of_iff ⟨λ H, subtype.eq H, subtype.mk.inj⟩)⟩
+
+theorem coeff_restriction {p : polynomial R} {n : ℕ} : ↑(coeff (restriction p) n) = coeff p n := rfl
+
+theorem coeff_restriction' {p : polynomial R} {n : ℕ} : (coeff (restriction p) n).1 = coeff p n := rfl
+
+theorem degree_restriction {p : polynomial R} : (restriction p).degree = p.degree := rfl
+
+theorem nat_degree_restriction {p : polynomial R} : (restriction p).nat_degree = p.nat_degree := rfl
+
+theorem monic_restriction {p : polynomial R} : monic (restriction p) ↔ monic p :=
+⟨λ H, congr_arg subtype.val H, λ H, subtype.eq H⟩
+
+theorem restriction_zero : restriction (0 : polynomial R) = 0 := rfl
+
+theorem restriction_one : restriction (1 : polynomial R) = 1 :=
+ext.2 $ λ i, subtype.eq $ by rw [coeff_restriction', coeff_one, coeff_one]; split_ifs; refl
+
+variables {S : Type v} [comm_ring S] {f : R → S} {x : S}
+
+theorem eval₂_restriction {p : polynomial R} :
+  eval₂ f x p = eval₂ (f ∘ subtype.val) x p.restriction :=
+rfl
+
+section base_change
+variables (p : polynomial R) (T : set R) [is_subring T]
+
+def base_change (hp : ↑p.frange ⊆ T) : polynomial T :=
```

Comment:
I don't think this should be called base change. The common understanding is that base change takes a ring hom `R → S`, and some object living over `R`, and attach an object over `S`. This is going in the other direction.

Reply:
I think some docstrings explaining the difference between `base_change` and `restriction` would be handy. `base_change` is actually what I would have expected `restriction` to be. I guess the same concepts come up in lots of places, so we should find a standard name.

Reply:
done, renamed to `to_subring`